### ISSUE TITLE
Allow option to skip examining existing entries

### ISSUE THREAD

**Issue by hfhchan** [jerickson314/sdscanner#7]
For some reason, whenever I scan a directory, e.g. /storage/external_SD, after the examination phase, the app keeps processing files under /storage/emulated/0/Android/data/.

I suspect that this is due to the app not having permission to access files under /storage/emulated/0/Android/data/, hence assuming they are removed and being processed.

I have millions of small picture files and the processing phase takes hours every time, so I often have to leave it on overnight. The process could definitely be sped up if it avoided reprocessing directories where I know there do not exist photos.

Maybe there could be an option to skip examining existing entries, and/or only process update/deleted files under the specified directory?

Anyhow, the app is great!  Saves me from the frustrating unreliable alternative of repeatedly plugging in and out my SD card and rebooting. Ugh.

**Comment by jerickson314** [comment]
There are actually three phases that happen.

In the first, when it says "Preparing initial list of files" it's recursively generating a list of all files that are actually present in the directory you specified. This usually happens very quickly, so there is no progress bar.

In the second, when it says "Querying media database" followed by "Examined ..." it's looking at the current contents of the media database to update the list of files that need scanning. For each updated or deleted file it finds, it makes sure it's in the list to scan. For each up-to-date file it finds, it makes sure it's NOT in the list to scan. The per-file time required for this process is easily an order of magnitude or two faster than for the third phase.

In the third and final phase, when it says "Processed ..." it's actually running the media scanner API on each file in the list remaining from the second phase. This is by far the slowest part of the process on a per-file basis.

In my earliest development versions of SD Scanner, I didn't have phase 2. This is what would happen if there were an option to not look at the existing entries. I found that it would take 5 or 10 minutes to scan my phone even when nothing had changed, compared to about 10 seconds for the three-phase process (due to not running the slow API on all those files). I don't think giving an option to skip phase 2 is a good idea, because in the common case for most users it will cause the scan to take a great deal longer.

What I'm guessing is happening in your case is that, as you guessed, a permissions problem causes the app to think those files have been deleted. This causes them to be added to the list in phase 2, so they get scanned in phase 3. So I went ahead and added an option in 1.10 to ignore updated/deleted files outside the scan directory.

I could theoretically speed up phase 2 by doing a proper SQL query on the media database in order to not even look at files outside the media directory. However, I've seen Android refer to the same SD card using multiple paths. If the path name SD Scanner uses differs from the one Android ends up writing to the media database, then phase 2 effectively won't happen due to the dumb string matching. To avoid that case, I think my current solution is probably best: reading the whole media database, and using getCanonicalFile() to see if the file we found in phase 1 is the same actual file as found in the media database.

Let me know if this version fixes your problem. The one possible wrinkle is that /storage/emulated/0 might be an alias of /storage/external_SD, in which case the current code will still detect /storage/emulated/0/Android/data as being inside /storage/external_SD.
